Mathlib/Analysis/Convex/Body.lean

```diff
@@ -0,0 +1,210 @@
+
+`ConvexBody V` is a module over the nonnegative reals (`NNReal`) and a pseudo-metric space.
+If `V` is a normed space, `ConvexBody V` is a metric space.
+
+## TODO
+
+- define positive convex bodies, requiring the interior to be nonempty
+- introduce support sets
+- Characterise the interaction of the distance with algebraic operations, eg
+  `dist (a • K) (a • L) = ‖a‖ * dist K L`, `dist (a +ᵥ K) (a +ᵥ L) = dist K L`
+
+## Tags
+
+convex, convex body
+-/
+
+
+open Pointwise
+
+open NNReal
+
+variable {V : Type _}
+
+/-- Let `V` be a real topological vector space. A subset of `V` is a convex body if and only if
+it is convex, compact, and nonempty.
+-/
+structure ConvexBody (V : Type _) [TopologicalSpace V] [AddCommMonoid V] [SMul ℝ V] where
+  carrier : Set V
+  convex' : Convex ℝ carrier
+  isCompact' : IsCompact carrier
+  nonempty' : carrier.Nonempty
+#align convex_body ConvexBody
+
+namespace ConvexBody
+
+section TVS
+
+variable [TopologicalSpace V] [AddCommGroup V] [Module ℝ V]
+
+instance : SetLike (ConvexBody V) V where
+  coe := ConvexBody.carrier
+  coe_injective' K L h := by
+    cases K
+    cases L
+    congr
+
+protected theorem convex (K : ConvexBody V) : Convex ℝ (K : Set V) :=
+  K.convex'
+#align convex_body.convex ConvexBody.convex
+
+protected theorem isCompact (K : ConvexBody V) : IsCompact (K : Set V) :=
+  K.isCompact'
+#align convex_body.is_compact ConvexBody.isCompact
+
+-- porting note: new theorem
+protected theorem isClosed [T2Space V] (K : ConvexBody V) : IsClosed (K : Set V) :=
+  K.isCompact.isClosed
+
+protected theorem nonempty (K : ConvexBody V) : (K : Set V).Nonempty :=
+  K.nonempty'
+#align convex_body.nonempty ConvexBody.nonempty
+
+@[ext]
+protected theorem ext {K L : ConvexBody V} (h : (K : Set V) = L) : K = L :=
+  SetLike.ext' h
+#align convex_body.ext ConvexBody.ext
+
+@[simp]
+theorem coe_mk (s : Set V) (h₁ h₂ h₃) : (mk s h₁ h₂ h₃ : Set V) = s :=
+  rfl
+#align convex_body.coe_mk ConvexBody.coe_mk
+
+section ContinuousAdd
+
+variable [ContinuousAdd V]
+
+-- we cannot write K + L to avoid reducibility issues with the set.has_add instance
+-- porting note: todo: is this^ still true?
+instance : Add (ConvexBody V) where
+  add K L :=
+    ⟨Set.image2 (· + ·) K L, K.convex.add L.convex, K.isCompact.add L.isCompact,
+      K.nonempty.add L.nonempty⟩
+
+instance : Zero (ConvexBody V) where
+  zero := ⟨0, convex_singleton 0, isCompact_singleton, Set.singleton_nonempty 0⟩
+
+instance : SMul ℕ (ConvexBody V) where
+  smul := nsmulRec
+
+-- porting note: todo: add `@[simp, norm_cast]`
+theorem coe_nsmul : ∀ (n : ℕ) (K : ConvexBody V), ↑(n • K) = n • (K : Set V)
+  | 0, _ => rfl
+  | (n + 1), K => congr_arg₂ (Set.image2 (· + ·)) rfl (coe_nsmul n K)
+
+instance : AddMonoid (ConvexBody V) :=
+  SetLike.coe_injective.addMonoid (↑) rfl (fun _ _ ↦ rfl) fun _ _ ↦ coe_nsmul _ _
+
+@[simp] -- porting note: todo: add `norm_cast`
+theorem coe_add (K L : ConvexBody V) : (↑(K + L) : Set V) = (K : Set V) + L :=
```

Comment:
I think it would be good to make these porting notes more explicit. I guess the idea is that `norm_cast` is not present in mathlib3 but should have been, but during the port it's better to stick to mathlib3 behavior and fix it later -- right? In this case how about "porting note: add `norm_cast`; we leave it out for now to reproduce mathlib3 behavior."

Reply:
If someone wants this PR to be merged before Wednesday, April 19: feel free to update the porting notes as suggested & merge. I have a few deadlines in a row right now...